components/family-studies/migrations/src/main/java/org/phenotips/studies/family/migrations/R71504PhenoTips3292DataMigration.java

```diff
@@ -0,0 +1,426 @@
+
+    private static final String CANCER_ID_LABEL = "id";
+
+    private static final String CANCER_LABEL_LABEL = "label";
+
+    private static final String AGE_AT_DIAGNOSIS = "ageAtDiagnosis";
+
+    private static final String NUMERIC_AGE_AT_DIAGNOSIS = "numericAgeAtDiagnosis";
+
+    private static final String NOTES = "notes";
+
+    private static final String AFFECTED = "affected";
+
+    private static final String QUALIFIERS = "qualifiers";
+
+    private static final String PRIMARY = "primary";
+
+    private static final Map<String, String> NAME_TO_ID_MAP = buildNameToIdMap();
+
+    /** Resolves unprefixed document names to the current wiki. */
+    @Inject
+    @Named("current")
+    private DocumentReferenceResolver<String> resolver;
+
+    /** Serializes the class name without the wiki prefix, to be used in the database query. */
+    @Inject
+    @Named("compactwiki")
+    private EntityReferenceSerializer<String> serializer;
+
+    /** Logging helper object. */
+    @Inject
+    private Logger logger;
+
+    @Override
+    public Object doInHibernate(final Session session) throws HibernateException, XWikiException
+    {
+        // Select all families.
+        final Query query = session.createQuery("select distinct o.name from BaseObject o where o.className= '"
+            + this.serializer.serialize(PEDIGREE_CLASS_REFERENCE) + "' and o.name <> 'PhenoTips.FamilyTemplate'");
+
+        @SuppressWarnings("unchecked")
+        List<String> docs = query.list();
+
+        this.logger.debug("Found {} documents", docs.size());
+
+        final XWikiContext context = getXWikiContext();
+        final XWiki xwiki = context.getWiki();
+        // Migrate each family.
+        docs.forEach(docName -> migrateFamily(docName, session, context, xwiki));
+        return null;
+    }
+
+    /**
+     * Migrates the family's cancer data.
+     *
+     * @param docName the name of the family {@link XWikiDocument}
+     * @param session the current {@link Session}
+     * @param context the {@link XWikiContext}
+     * @param xwiki the {@link XWiki}
+     */
+    private void migrateFamily(
+        @Nonnull final String docName,
+        @Nonnull final Session session,
+        @Nonnull final XWikiContext context,
+        @Nonnull final XWiki xwiki)
+    {
+        try {
+            final XWikiDocument xDocument = xwiki.getDocument(this.resolver.resolve(docName), context);
+
+            if (xDocument != null) {
+                this.logger.debug("Updating pedigree for document {}.", docName);
+                final BaseObject pedigreeXObject = xDocument.getXObject(PEDIGREE_CLASS_REFERENCE);
+                if (pedigreeXObject != null) {
+                    updatePedigree(xDocument, pedigreeXObject, context, session);
+                }
+            }
+        } catch (final JSONException ex) {
+            this.logger.error("Error updating JSON for document {} : [{}]", docName, ex.getMessage());
```

Comment:
In theory this may be caused by reading existing JSON (not writing new stuff), so the error message is a bit misleading - though the end result is the same, the data does not get updated, so I think this is OK